crates/swc_common/src/plugin.rs

```diff
@@ -8,38 +8,58 @@ use crate::{syntax_pos::Mark, SyntaxContext};
 use anyhow::Error;
 use std::any::type_name;
 
-/// Serialize ast to pass into plugins. This is being called from host side.
-/// TODO: https://github.com/swc-project/swc/issues/3167
-#[cfg(feature = "plugin-base")]
-pub fn serialize_for_plugin<T>(t: &T) -> Result<rkyv::AlignedVec, Error>
+/// A struct wraps internal representation of serialized Program to be passed
+/// across plugin's host to guest. Plugin author should not rely on specific
+/// details serialized byte format struct contains: it is strictly
+/// implementation detail which can change anytime.
+#[cfg(any(feature = "plugin-mode", feature = "plugin-rt"))]
+pub struct SerializedProgram(pub rkyv::AlignedVec);
+
+impl SerializedProgram {
+    #[cfg(any(feature = "plugin-mode", feature = "plugin-rt"))]
+    pub fn new(bytes: Vec<u8>, len: i32) -> SerializedProgram {
+        let mut vec = rkyv::AlignedVec::with_capacity(
+            len.try_into()
+                .expect("Cannot determine size of the serialized bytes"),
+        );
+        vec.extend_from_slice(&bytes);
+        SerializedProgram(vec)
+    }
+}
+
+/// Serialize given Program into raw bytes, can be copied into plugin's memory
+/// spaces.
+#[cfg(any(feature = "plugin-rt", feature = "plugin-mode"))]
+pub fn serialize_into_bytes<T>(t: &T) -> Result<SerializedProgram, Error>
 where
     T: rkyv::Serialize<rkyv::ser::serializers::AllocSerializer<512>>,
 {
-    rkyv::to_bytes::<_, 512>(t).map_err(|err| match err {
-        rkyv::ser::serializers::CompositeSerializerError::SerializerError(e) => e.into(),
-        rkyv::ser::serializers::CompositeSerializerError::ScratchSpaceError(e) => {
-            Error::msg("AllocScratchError")
-        }
-        rkyv::ser::serializers::CompositeSerializerError::SharedError(e) => {
-            Error::msg("SharedSerializeMapError")
-        }
-    })
+    rkyv::to_bytes::<_, 512>(t)
+        .map(|v| SerializedProgram(v))
+        .map_err(|err| match err {
+            rkyv::ser::serializers::CompositeSerializerError::SerializerError(e) => e.into(),
+            rkyv::ser::serializers::CompositeSerializerError::ScratchSpaceError(e) => {
+                Error::msg("AllocScratchError")
+            }
+            rkyv::ser::serializers::CompositeSerializerError::SharedError(e) => {
+                Error::msg("SharedSerializeMapError")
+            }
+        })
 }
 
-/// TODO: https://github.com/swc-project/swc/issues/3167
-#[cfg(feature = "plugin-mode")]
-pub fn deserialize_for_plugin<T>(bytes: &[u8]) -> Result<T, Error>
```

Comment:
`deserialize_for_plugin` exists because we need to serialize/deserialize diagnostics generated by plugins.
This should not be renamed, and I don't think we have to make a breaking change here.

Reply:
It was bikeshedding to have somewhat better clarity, and should be fine to revert.